types/test.ts

```diff
@@ -0,0 +1,203 @@
+      cache: new Map(),
+      datetime: "value",
+      appendSessionToken: true,
+      allHeaders: true,
+      singleEncode: true,
+    },
+    cache: "default",
+    credentials: "same-origin",
+    integrity: "value",
+    keepalive: true,
+    mode: "cors",
+    redirect: "follow",
+    referrer: "value",
+    referrerPolicy: "same-origin",
+    signal: new AbortSignal(),
+    window: null,
+  };
+}
+
+function AwsV4SignerTest() {
+  function constructor() {
+    // $ExpectError
+    new AwsV4Signer();
+
+    new AwsV4Signer({
+      url: "value",
+      accessKeyId: "value",
+      secretAccessKey: "value",
+      retries: 1, // $ExpectError
+    });
+
+    new AwsV4Signer({
+      url: "value",
+      accessKeyId: "value",
+      secretAccessKey: "value",
+      initRetryMs: 1, // $ExpectError
+    });
+
+    new AwsV4Signer({
+      url: "value",
+      accessKeyId: "value",
+      secretAccessKey: "value",
+      body: new Blob(), // $ExpectError
+    });
+
+    new AwsV4Signer({
+      url: "value",
+      accessKeyId: "value",
+      secretAccessKey: "value",
+      body: new FormData(), // $ExpectError
+    });
+
+    new AwsV4Signer({
+      url: "value",
+      accessKeyId: "value",
+      secretAccessKey: "value",
+      body: new URLSearchParams(), // $ExpectError
+    });
+
+    new AwsV4Signer({
+      url: "value",
+      accessKeyId: "value",
+      secretAccessKey: "value",
+      body: new ReadableStream(), // $ExpectError
+    });
+
+    const min = new AwsV4Signer({
+      url: "value",
+      accessKeyId: "value",
+      secretAccessKey: "value",
+    });
+
+    const all = new AwsV4Signer({
+      url: "value",
+      accessKeyId: "value",
+      secretAccessKey: "value",
+      sessionToken: "value",
+      method: "value",
+      headers: {},
+      body: "value",
+      signQuery: true,
+      service: "value",
+      region: "value",
+      cache: new Map(),
+      datetime: "value",
+      appendSessionToken: true,
+      allHeaders: true,
+      singleEncode: true,
+    });
+  }
+
+  function authHeader(signer: AwsV4Signer) {
+    // $ExpectType Promise<string>
+    signer.authHeader();
+  }
+
+  function sign() {
+    // $ExpectType Promise<AwsRequestInfo<undefined>>
+    new AwsV4Signer({
+      url: "value",
+      accessKeyId: "value",
+      secretAccessKey: "value",
+    }).sign();
+
+    // $ExpectType Promise<AwsRequestInfo<"value">>
+    new AwsV4Signer({
+      url: "value",
+      accessKeyId: "value",
+      secretAccessKey: "value",
+      body: "value",
+    }).sign();
+
+    // $ExpectType Promise<AwsRequestInfo<ArrayBuffer>>
+    new AwsV4Signer({
+      url: "value",
+      accessKeyId: "value",
+      secretAccessKey: "value",
+      body: new ArrayBuffer(1),
+    }).sign();
+
+    // $ExpectType Promise<AwsRequestInfo<Uint8Array>>
+    new AwsV4Signer({
+      url: "value",
+      accessKeyId: "value",
+      secretAccessKey: "value",
+      body: new Uint8Array(),
+    }).sign();
```

Comment:
Previous types don't provide generics support for getting a properly types `body`.